### PR TITLE
Added logics for restoration of multinode cluster from bucket.

### PR DESCRIPTION
**Pull request by abdasgupta**
**What this PR does / why we need it**:
This PR completes the bootstrap scenario of a multi node cluster
**Which issue(s) this PR fixes**:
Fixes #
Fixes the issue of bootstrapping of ETCD members of a multi node cluster as individual member while restoration
**Special notes for your reviewer**:

**Release note**:

```improvement operator

```

### CODE REVIEW

**Review by ishan16696 — CHANGES_REQUESTED** [review]
Can you also take a look at integration tests as tests got failed?

File: pkg/snapshot/restorer/restorer.go [hunk 1]
```diff
@@ -713,3 +731,53 @@ func verifySnapshotRevision(clientKV client.KVCloser, snap *brtypes.Snapshot) er
...
+
+func (r *Restorer) isLeader(etcdConnectionConfig *brtypes.EtcdConnectionConfig) bool {
+	factory := etcdutil.NewFactory(*etcdConnectionConfig)
+	clientMaintenance, err := factory.NewMaintenance()
+	if err != nil {
+		r.logger.Errorf("Could not create etcd client while determining if the embedded etcd is leader: %v", err)
+		return false
+	}
+
+	response, err := clientMaintenance.Status(context.Background(), defaultEtcdConnectionEndpoint)
```
Comment (aaronfern) [inline pkg/snapshot/restorer/restorer.go:776]:
```suggestion
	response, err := clientMaintenance.Status(context.Background(), etcdConnectionConfig.Endpoints[0])
```
Uses the default endpoint even if one is explicitly set

File: pkg/snapshot/restorer/restorer.go [hunk 2]
```diff
@@ -104,20 +106,36 @@ func (r *Restorer) Restore(ro brtypes.RestoreOptions) (*embed.Etcd, error) {
...
-	clientFactory := etcdutil.NewClientFactory(ro.NewClientFactory, brtypes.EtcdConnectionConfig{
+	cfg := &brtypes.EtcdConnectionConfig{
 		MaxCallSendMsgSize: ro.Config.MaxCallSendMsgSize,
 		Endpoints:          []string{e.Clients[0].Addr().String()},
 		InsecureTransport:  true,
-	})
+	}
+	if ro.Config.ClientTLSInfo != nil {
+		cfg.Endpoints = []string{"https://etcd-main-local:2379"}
+		cfg.CaFile = "/var/etcd/ssl/client/client/ca/ca.crt"
```
Comment (aaronfern) [inline pkg/snapshot/restorer/restorer.go:116]:
```suggestion
		cfg.CaFile = "/var/etcd/ssl/client/client/ca.crt"
```

**Review by aaronfern — CHANGES_REQUESTED** [review]

**Comment by aaronfern** [comment]
I tried running the restoration scenario for single node etcd's as well as multi node etcd's, with both TLS enables as well as disabled.
Running fine in all cases

File: pkg/snapshot/restorer/restorer.go (hunk 1 above)
Comment (abdasgupta) [reply pkg/snapshot/restorer/restorer.go]: Done

File: pkg/snapshot/restorer/restorer.go [hunk 3]
```diff
@@ -713,3 +724,67 @@ func verifySnapshotRevision(clientKV client.KVCloser, snap *brtypes.Snapshot) er
...
+	}
+
+	factory := etcdutil.NewFactory(*etcdConnectionConfig)
+	clientMaintenance, err := factory.NewMaintenance()
+	if err != nil {
+		r.logger.Errorf("Could not create etcd client while determining if the embedded etcd is leader: %v", err)
+		return false
+	}
+
+	response, err := clientMaintenance.Status(context.Background(), defaultEtcdConnectionEndpoint)
```
[the remark this answers is not in the document]
Comment (abdasgupta) [reply pkg/snapshot/restorer/restorer.go]: Done

File: pkg/snapshot/restorer/restorer.go [hunk 4]
```diff
@@ -56,8 +57,10 @@ import (
 )
 
 const (
-	tmpDir                  = "/tmp"
-	tmpEventsDataFilePrefix = "etcd-restore-"
+	tmpDir                               = "/tmp"
+	tmpEventsDataFilePrefix              = "etcd-restore-"
+	defaultEtcdConnectionEndpoint        = "0.0.0.0:2379"
+	NoLeaderState                 uint64 = 0
```
[the remark this answers is not in the document]
Comment (abdasgupta) [reply pkg/snapshot/restorer/restorer.go]: Done

File: pkg/server/backuprestoreserver.go [hunk 5]
```diff
@@ -88,20 +94,150 @@ func NewBackupRestoreServer(logger *logrus.Logger, config *BackupRestoreComponen
 
 // Run starts the backup restore server.
 func (b *BackupRestoreServer) Run(ctx context.Context) error {
-	clusterURLsMap, err := types.NewURLsMap(b.config.RestorationConfig.InitialCluster)
+	inputFileName := "/var/etcd/config/etcd.conf.yaml"
+	configYML, err := ioutil.ReadFile(inputFileName)
```
[the remark this answers is not in the document]
Comment (abdasgupta) [reply pkg/server/backuprestoreserver.go]: Done

File: pkg/snapshot/restorer/restorer_test.go [hunk 6]
```diff
@@ -50,7 +50,7 @@ const (
 	allSnapsInV2
 )
 
-var _ = Describe("Running Restorer", func() {
+var _ = FDescribe("Running Restorer", func() {
```
[the remark this answers is not in the document]
Comment (abdasgupta) [reply pkg/snapshot/restorer/restorer_test.go]: Done

File: pkg/snapshot/restorer/restorer.go [hunk 7]
```diff
@@ -713,3 +724,67 @@ func verifySnapshotRevision(clientKV client.KVCloser, snap *brtypes.Snapshot) er
...
+		if err := verifySnapshotRevisionForAll(clientKV, lastDeltaSnap); err != nil {
+			r.logger.Errorf("All delta snapshots are not yet applied for %v because : %v", v[0].String(), err)
+			return false
+		}
+	}
+	return true
+}
+
+func verifySnapshotRevisionForAll(clientKV client.KVCloser, snap *brtypes.Snapshot) error {
+	ctx := context.TODO()
```
[the remark this answers is not in the document]
Comment (abdasgupta) [reply pkg/snapshot/restorer/restorer.go]: Done

**Comment by aaronfern** [comment]
I noticed that there was a regression with compaction with the changes from this PR
The issue is [here](https://github.com/gardener/etcd-backup-restore/blob/722eea3155e604fe222150819c601840f34619e0/pkg/types/restorer.go#L163). Since this PR changes the peer port used by the embedded etcd from `2380` to `2480`, this needs to be updated as well

**Comment by ishan16696** [comment]
Closing this PR as now it not required now and it has become quite old now.
/close

**Pull request closed without merging**